### ISSUE TITLE
[HOLD for payment 2025-01-21] [$250] Track- Error appears when Submit expense to someone

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#45086]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.5-3
**Reproducible in staging?:** Y
**Reproducible in production?:** Y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4704990&group_by=cases:section_id&group_order=asc&group_id=309130
**Email or phone of affected tester (no customers):** gocemate+a573@gmail.com

## Action Performed:
1. Login with new account
2. Go to Self DM>  Track expense
3. Submit the expense to someone
4. Wait for a minute

## Expected Result:
There should be no errors and submitted expense should be present on LHN and report

## Actual Result:
Red dot and error appears after Submit expense to someone

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/e719bca2-468b-491e-a7e0-3b5721e2c76f

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01408041e16e8fde12</li>
        <li>Upwork Job ID: 1815525298284767197</li>
        <li>Last Price Increase: 2024-08-08</li>
<li>Automatic offers: </li>
<ul>
<li>CyberAndrii | Contributor | 105455137</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @alexpensify</details>

**Comment by lanitochka17** [comment]
@alexpensify  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by alexpensify** [comment]
This one is on my testing list. I'll review it soon.

**Comment by alexpensify** [comment]
No update, I'll test soon

**Comment by alexpensify** [comment]
Other GHs have been a priority, I'll review soon

**Comment by alexpensify** [comment]
I was able to replicate the experience in my Android application 

![image](https://github.com/user-attachments/assets/579ff65a-7018-45b4-b5ab-8b1e85dc0fed)

![image](https://github.com/user-attachments/assets/a739c91d-7777-4787-b5c9-57a3e9240c9d)

**Comment by alexpensify** [comment]
We need proposals here.

**Comment by alexpensify** [comment]
Open for proposals here

**Comment by alexpensify** [comment]
Open for proposals here

**Comment by pasyukevich** [comment]
Hi, I'm Eugene from Callstack - expert contributor group. I’d like to work on this issue.

**Comment by alexpensify** [comment]
Thanks @pasyukevich! I've assigned this one to you.

**Comment by alexpensify** [comment]
@pasyukevich any update on when the PR will be ready for review? Thanks!

**Comment by pasyukevich** [comment]
I was able to reproduce the issue on Android. 
I am looking for the root cause and don't have an ETA for the PR.

Do you also mean that here we do not need to have the proposal step and I need to create the PR?

**Comment by jacobkim9881** [comment]
@alexpensify Hi. Is the issue still open to proposal? If it is, I would like to investigate this issue.

**Comment by alexpensify** [comment]
@pasyukevich - No, you would create the PR only. 

--- 

@jacobkim9881 this one is not open right now. Thanks for your interest.

**Comment by alexpensify** [comment]
Moving to Weekly while @pasyukevich works on this one.

**Comment by trjExpensify** [comment]
@thienlnam any context on why we add this "changed the expense" system message in this flow? 🤔

**Comment by thienlnam** [comment]
We add a MODIFIEDEXPENSE reportAction when the tracked expense is submitted to someone for historical context. However, I'm not sure that message is ideal for this scenario. I think we just need to change the copy here maybe to something like 'submitted tracked expense'?

**Comment by trjExpensify** [comment]
Why did we decide to add a`MODIFIEDEXPENSE` reportAction for historical context of an expense being submitted not modified? 🤔 

>I think we just need to change the copy here maybe to something like 'submitted tracked expense'?

Even this doesn't seem necessary. If I go to `Submit expense` > Jack, we don't add the "submitted $x" reportAction in the iouReport. So why do we need to do that here? The only time we use that reportAction is when a workspace has delayed submission enabled and the expense report is actually submitted (be that manually, or via harvesting). 

iouReports are the equivalent of `instant` submit, where the concept of having to submit a report doesn't exist. So I think we remove this reportAction altogether.

**Comment by thienlnam** [comment]
This was a tracked expense, and then when it was submitted it became a reimbursable expense which is why it has a modified reportAction. I feel like there needs to be some indication that something happened to this tracked expense

**Comment by JmillsExpensify** [comment]
But they don't know one way or the other right? Like a tracked expense is basically secretly non-reimbursable.

**Comment by trjExpensify** [comment]
^^ I agree with that

**Comment by thienlnam** [comment]
I don't think it matters to the person that is getting the expense submitted to them, but for the person that created the tracked expense it might be nice to see that this tracked expense has been updated and is now being submitted to someone. 

Though I don't feel that strongly so if you want we can remove the reportAction that gets added when the expense is moved

**Comment by trjExpensify** [comment]
I think redirecting them to the DM and seeing this preview that reads "<person you just submitted to> owes:" is fine, personally. 

<img width="330" alt="image" src="https://github.com/user-attachments/assets/fce8a192-2b32-4db2-bdbf-a93f5d347841">

**Comment by alexpensify** [comment]
@pasyukevich—The team has replied with feedback. Do you have enough information to continue with the PR, or should we have a larger discussion in the Open Source Slack room? Thanks for the update!

**Comment by pasyukevich** [comment]
@alexpensify    I am  not sure, still investigating the exact reason - A larger discussion will be helpful 

I've found on the android a few different flows of this problem: 

 1. Unsynced state with BE - when we send straight after the creation of the `convertTrackedExpenseToRequest` and receive a response with the  not found state  (this is reproducible also on other platforms, floating bug)
 2. multiple requests of `convertTrackedExpenseToRequest` - we send `convertTrackedExpenseToRequest` and receive a response with status 200, but straight after it sends one more and receives 401 (as the previous request has already converted it) - error at the app as a result
 
 I'm focused on the 2nd case as it reproduces only on the Android constantly (1st - floating problem)  
 
 It was hard to investigate the exact flow, since on the js level we always have the same error - [TypeError: Network request failed] (even if `convertTrackedExpenseToRequest` succeeded on the native side )
 
I've configured the logs from the native Android and network sides with exact errors - it is much easier to investigate the root cause

**Comment by alexpensify** [comment]
Ok, I'll start one tomorrow.

**Comment by CyberAndrii** [comment]
Coming from https://github.com/Expensify/App/issues/48238. This error is caused by the [`Receipt.type`](https://github.com/Expensify/App/blob/fc849dda35628b8c6c8d6edfa66dd31b257e8e7a/src/types/onyx/Transaction.ts#L138) property being `undefined`. We had a similar issue before: https://github.com/Expensify/App/issues/26940#issuecomment-1740445802.

`receipt` must be either `undefined` or has a value assigned to the `type` property. You can quickly fix it with this change:
```diff
diff --git a/src/libs/actions/IOU.ts b/src/libs/actions/IOU.ts
index 29d48173779..4414d8221c7 100644
--- a/src/libs/actions/IOU.ts
+++ b/src/libs/actions/IOU.ts
@@ -3553,7 +3553,7 @@ function requestMoney(
                 createdChatReportActionID,
                 createdIOUReportActionID,
                 reportPreviewReportActionID: reportPreviewAction.reportActionID,
-                receipt,
+                receipt: receipt && Object.keys(receipt).length === 0 ? undefined : receipt,
                 receiptState: receipt?.state,
                 category,
                 tag,
```

This will only partially fix the issue as there are more places where a file is passed but the type is missing.

**Comment by alexpensify** [comment]
I'm catching up from being OOO last week and will start the required discussion soon.

**Comment by alexpensify** [comment]
No update

**Comment by alexpensify** [comment]
Other GitHubs have been a priority, I'll get to this one soon.

**Comment by alexpensify** [comment]
I've been sick this week and haven't been able to review this GitHub. There is no major update here; I'll try to carry on next week.

**Comment by pasyukevich** [comment]
I've rechecked the issue - it is still reproducible with the same logs

**Comment by alexpensify** [comment]
No update

**Comment by alexpensify** [comment]
@Pujan92 - I should have put this one on hold for https://github.com/Expensify/App/issues/48238, but that one went into production recently. Can you please confirm what's left, as this appears to be a multilevel issue, or should we close this GH? Thanks!

**Comment by alexpensify** [comment]
@Pujan92 any update here? Thanks!

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Tuesday, October 22, 2024, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If it can wait, I'll review it when I return online.

**Comment by Pujan92** [comment]
I see the same error as @pasyukevich mentioned 

[Network] API request error: Connection interruption likely - "Network request failed"

@pasyukevich, is there any update about why this error is being thrown?

**Comment by flaviadefaria** [comment]
@al, this makes more sense as part of #convert since this is part of the bottom-up adoption workflow.

**Comment by pasyukevich** [comment]
@Pujan92 There are no updates, the latest one - https://github.com/Expensify/App/issues/45086#issuecomment-2293900646

**Comment by CyberAndrii** [comment]
> [Network] API request error: Connection interruption likely - "Network request failed"

See https://github.com/Expensify/App/issues/45086#issuecomment-2323920107 for why this happens

**Comment by alexpensify** [comment]
@Pujan92 any thoughts about the feedback shared so far? Thanks!

**Comment by alexpensify** [comment]
Update: Waiting for proposals or a proposal update

**Comment by alexpensify** [comment]
Update: Same as last week

**Comment by alexpensify** [comment]
Update: waiting for proposals here

**Comment by alexpensify** [comment]
There are no proposals yet. I'll need to start shopping around for help.

**Comment by CyberAndrii** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Track - Error appears when Submit expense to someone

### What is the root cause of that problem?

The `receipt` object is passed directly as a parameter to `API.write` [here](https://github.com/Expensify/App/blob/d56dfeca08a93d9b3a131c8ba17e4783758dad0e/src/libs/actions/IOU.ts#L3575). The issue is that it gets incorrectly serialized as `[object Object]`.

Screenshot from the Network tab:

![image](https://github.com/user-attachments/assets/9f0b1d73-e3f1-490c-840b-8ebf4139207c)

This error only appears on Android devices due to more assertive implementation of `XMLHttpRequest`. To be specific the error happens on this line

https://github.com/facebook/react-native/blob/a28867f9524beeabb3cb939446fe825cac124c58/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L753-L754 

### What changes do you think we should make in order to solve the problem?

We need to convert the `receipt` object to a JSON string first

https://github.com/Expensify/App/blob/d56dfeca08a93d9b3a131c8ba17e4783758dad0e/src/libs/actions/IOU.ts#L3575

```ts
receipt: JSON.stringify(receipt),
```

There are a few other places where this fix will be needed

https://github.com/Expensify/App/blob/d56dfeca08a93d9b3a131c8ba17e4783758dad0e/src/libs/actions/IOU.ts#L3704

https://github.com/Expensify/App/blob/d56dfeca08a93d9b3a131c8ba17e4783758dad0e/src/libs/actions/IOU.ts#L3819

https://github.com/Expensify/App/blob/d56dfeca08a93d9b3a131c8ba17e4783758dad0e/src/libs/actions/IOU.ts#L4087

https://github.com/Expensify/App/blob/d56dfeca08a93d9b3a131c8ba17e4783758dad0e/src/libs/actions/IOU.ts#L5065

I also suggest printing a console message from the [`xhr`](https://github.com/Expensify/App/blob/94ab9cd0e37faa5facdf858952eab42a3169df7f/src/libs/HttpUtils.ts#L160-L173) function to make such errors easier to spot.

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?

Add unit tests to verify that object parameters are converted to JSON strings before being sent to the API.

### What alternative solutions did you explore? (Optional)

Instead of updating each instance manually, we could handle this globally for all requests

https://github.com/Expensify/App/blob/94ab9cd0e37faa5facdf858952eab42a3169df7f/src/libs/HttpUtils.ts#L160-L173

```diff
 function xhr(command: string, data: Record<string, unknown>, type: RequestType = CONST.NETWORK.METHOD.POST, shouldUseSecure = false): Promise<Response> {
     const formData = new FormData();
     Object.keys(data).forEach((key) => {
-        if (typeof data[key] === 'undefined') {
+        const value = data[key];
+        if (value === undefined) {
             return;
         }
-        formData.append(key, data[key] as string | Blob);
+        if (typeof value === 'object' && !isBlob(value)) {
+            formData.append(key, JSON.stringify(value));
+            return;
+        }
+        formData.append(key, value as string | Blob);
     });
  
     const url = ApiUtils.getCommandURL({shouldUseSecure, command}); 
  
     const abortSignalController = data.canCancel ? abortControllerMap.get(command as AbortCommand) ?? abortControllerMap.get(ABORT_COMMANDS.All) : undefined; 
     return processHTTPRequest(url, type, formData, abortSignalController?.signal); 
 }

+/**
+ * Returns true if the object is a "blob", which in React Native just means an object with a `uri` attribute.
+ * Optionally, it can also have a `name` and `type` attribute to specify filename and content type.
+ */
+function isBlob(object: {} | null) {
+    return !!object && typeof object === 'object' && !Array.isArray(object) && 'uri' in object;
+}
```

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Wednesday, December 18, 2024, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If the inquiry can wait, I'll review it when I return online. 

--- 

When I'm back online, I'll start asking for help.

**Comment by CyberAndrii** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Track - Error appears when Submit expense to someone

### What is the root cause of that problem?

- When a new "Scan" expense is created, we send a `File` object, which gets serialized as binary - no errors

  ![image](https://github.com/user-attachments/assets/fca5b8a0-2090-4972-a0d5-64cacfe5566b)

  https://github.com/Expensify/App/blob/f0461fbfd9d2d0da938365749657f6487c591739/src/pages/iou/request/step/IOURequestStepConfirmation.tsx#L209-L213

- When a new "Manual" expense is created, `receipt` is `undefined` so nothing is sent  - no errors

- When a new "Distance" expense is created, we send a regular object `{source: "...", state: "..."}` created [here](https://github.com/Expensify/App/blob/b8579b1ab0445c32bea41486b28bb78a240b1454/src/libs/actions/IOU.ts#L3945-L3946) - causes an error

- When a tracked expense is submitted to someone, categorized, or shared with an accountant, we send a regular object `{receiptID: ..., source: "...", state: "..."}` or `{}` returned from the backend - causes an error

  https://github.com/Expensify/App/blob/f0461fbfd9d2d0da938365749657f6487c591739/src/pages/iou/request/step/IOURequestStepConfirmation.tsx#L204-L207

The reason is that `XMLHttpRequest`/`FormData` only allows sending
- on web: either a `string` or a `Blob`. `File` extends `Blob` so it also works.
- on native: either a `string` or an object that satisfies `{name?: string, type?: string, uri: string}` _([link to the source](https://github.com/facebook/react-native/blob/18243d07e41fb030bb09c760b96a22d9fdc4f948/packages/react-native/Libraries/Network/FormData.js#L13))_. `File` satisfies that so it also works.
  
All other types
- on web: will be converted by `String(value)`, producing a string like `[object Object]`.
  ![image](https://github.com/user-attachments/assets/c3d0a8f3-bf26-42ef-92b9-cc984ebe322d)
- on android:
    - if `typeof value === 'object' && !Array.isArray(value)`: will throw an error _([link to the source](https://github.com/facebook/react-native/blob/a28867f9524beeabb3cb939446fe825cac124c58/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L753-L754))_ - that's the reason of this issue
    - otherwise: will be converted to a string _([link to the source](https://github.com/facebook/react-native/blob/18243d07e41fb030bb09c760b96a22d9fdc4f948/packages/react-native/Libraries/Network/FormData.js#L103))_
- on ios: not sure, I didn't test

### What changes do you think we should make in order to solve the problem?

For [`TrackExpense`](https://github.com/Expensify/App/blob/64d6c3875cf052a51b88e31ebe6eee3f0a230985/src/libs/actions/IOU.ts#L4085) and [`RequestMoney`](https://github.com/Expensify/App/blob/64d6c3875cf052a51b88e31ebe6eee3f0a230985/src/libs/actions/IOU.ts#L3822): only send the `receipt` if it's a file

```diff
-receipt,
+receipt: receipt instanceof File || receipt instanceof Blob ? receipt : undefined,

// optionally also

-receiptState,
+receiptState: receipt instanceof File || receipt instanceof Blob ? receipt?.state : undefined,
```

For [`ConvertTrackedExpenseToRequest`](https://github.com/Expensify/App/blob/64d6c3875cf052a51b88e31ebe6eee3f0a230985/src/libs/actions/IOU.ts#L3582), [`CategorizeTrackedExpense`](https://github.com/Expensify/App/blob/64d6c3875cf052a51b88e31ebe6eee3f0a230985/src/libs/actions/IOU.ts#L3623), and [`ShareTrackedExpense`](https://github.com/Expensify/App/blob/64d6c3875cf052a51b88e31ebe6eee3f0a230985/src/libs/actions/IOU.ts#L3709): we could either do the same as above or just never send the `receipt` as it doesn't make much sense for these requests to upload it. From my testing, it looks like the BE doesn't even expect it

```diff
-receipt,
```

For [`CategorizeTrackedExpense`](https://github.com/Expensify/App/blob/64d6c3875cf052a51b88e31ebe6eee3f0a230985/src/libs/actions/IOU.ts#L3623) we also need to remove `onyxData` and `linkedTrackedExpenseReportAction` as they were accidentally added [here](https://github.com/Expensify/App/pull/52221/files#r1880255805):

```diff

 const parameters = {
-    onyxData,
     ...reportInformation,
     ...policyParams,
     ...transactionParams,
+    receipt: undefined,
+    linkedTrackedExpenseReportAction: undefined,
     modifiedExpenseReportActionID,
     policyExpenseChatReportID: createdWorkspaceParams?.expenseChatReportID,
     policyExpenseCreatedReportActionID: createdWorkspaceParams?.expenseCreatedReportActionID,
     adminsChatReportID: createdWorkspaceParams?.adminsChatReportID,
     adminsCreatedReportActionID: createdWorkspaceParams?.adminsCreatedReportActionID,
 };
```

Test branch: https://github.com/CyberAndrii/ExpensifyApp/commit/6855d521b18b0c2cddd7b788624b8ae9d9feedcd

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?

Unit tests for the above functions to ensure that only objects of type `File` and `Blob` are sent to the backend (and never `{ ... }`).

I also suggest printing a console warning message from the [`xhr`](https://github.com/Expensify/App/blob/94ab9cd0e37faa5facdf858952eab42a3169df7f/src/libs/HttpUtils.ts#L160-L173) function to make such errors easier to spot.

**Comment by alexpensify** [comment]
@Pujan92 can you please review the latest update? Thanks!

**Comment by Pujan92** [comment]
@CyberAndrii's [proposal](https://github.com/Expensify/App/issues/45086#issuecomment-2546639916) of not passing the receipt param correctly looks promising to me which leads to the formdata unsupported type issue. Also, I would like to know the thoughts of @pasyukevich 

🎀👀🎀 C+ reviewed

**Comment by blimpich** [comment]
I agree with @Pujan92. Proposal looks promising. @pasyukevich could you weigh in when you get a chance?

**Comment by pasyukevich** [comment]
The proposal looks very promising to me as well!

Before removing the parameter, I think it is needed to verify on the API side to ensure that the `receipt` parameter is not required in any scenario for `ConvertTrackedExpenseToRequest`, `CategorizeTrackedExpense`, and `ShareTrackedExpense`.

**Comment by alexpensify** [comment]
I'm moving to Weekly since things will be a little slower this week. @blimpich, when you are back online, can you please review this one? Thanks!

**Comment by blimpich** [comment]
> The proposal looks very promising to me as well!
> 
> Before removing the parameter, I think it is needed to verify on the API side to ensure that the `receipt` parameter is not required in any scenario for `ConvertTrackedExpenseToRequest`, `CategorizeTrackedExpense`, and `ShareTrackedExpense`.

Can confirm it is not required at all for `ConvertTrackedExpenseToRequest`, and it's optional for `CategorizeTrackedExpense` and `ShareTrackedExpense`. Going to assign @CyberAndrii to this 👍

One question @CyberAndrii. Your proposal says this breaks for the tracked case and for the distance request case correct? We definitely should not be sending anything for the receipt file in the distance request, so can you fix it for both the tracked case and the distance case?

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Friday, January 3, 2025, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If the inquiry can wait, I'll review it when I return online.

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Friday, January 3, 2025, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If the inquiry can wait, I'll review it when I return online.

**Comment by blimpich** [comment]
@CyberAndrii @Pujan92 looks like we had to revert? What are the next steps here? Are we waiting on https://github.com/Expensify/App/pull/54358?

**Comment by CyberAndrii** [comment]
@blimpich All good. It was a partial revert and re-implementation in a different way.

But QA is having trouble testing this due to an unrelated issue. Commented [here](https://github.com/Expensify/App/pull/54628#issuecomment-2588251566).

https://github.com/Expensify/App/pull/54358 is also unrelated. They just wanted to confirm if the merge conflicts were resolved correctly.

**Comment by alexpensify** [comment]
I'll work on the payment process tomorrow.

**Comment by garrettmknight** [comment]
$250 approved for @Pujan92
